manticore/platforms/evm_world_state.py

```diff
@@ -0,0 +1,445 @@
+
+
+_endpoints: Dict[str, Endpoint] = {}
+
+
+def _web3_address(address: int) -> ChecksumAddress:
+    return Web3.toChecksumAddress("0x%0.40x" % address)
+
+
+# sam.moelius: Notes:
+#
+# 1. https://github.com/ethereum/wiki/wiki/JSON-RPC lists the kinds of information that an Ethereum
+#    node can provide over JSON RPC.
+#
+# 2. The accounts and get_storage methods do not make sense when using JSON RPC.  IMHO, we should
+#    program to the least common denominator.  In that sense, we should see whether the accounts and
+#    get_storage methods could be eliminated.
+
+
+class RemoteWorldState(WorldState):
+    def __init__(self, url: str):
+        actual = urlparse(url)
+        expected = ParseResult(scheme="", netloc="", path=url, params="", query="", fragment="")
+        if actual != expected:
+            raise EthereumError("URL must be of the form 'IP:PORT': " + url)
+        self._url = url
+
+    def _web3(self) -> Web3:
+        # sam.moelius: Force WebsocketProvider.__init__ to call _get_threaded_loop.  The existing
+        # "threaded loop" could be leftover from a fork, in which case it will not work.
+        Web3.WebsocketProvider._loop = None
+        web3 = Web3(Web3.WebsocketProvider(URI("ws://" + self._url)))
+        blocknumber = web3.eth.blockNumber
+        endpoint = _endpoints.get(self._url)
+        if endpoint is None:
+            endpoint = Endpoint(blocknumber, False)
+            _endpoints[self._url] = endpoint
+            logger.info("Connected to %s (blocknumber = %d)", self._url, blocknumber)
+        if endpoint.blocknumber != blocknumber:
+            if not endpoint.warned:
+                logger.warning(
+                    "%s blocknumber has changed: %d != %d---someone is using the endpoint besides us",
+                    self._url,
+                    endpoint.blocknumber,
+                    blocknumber,
+                )
+                endpoint.warned = True
+        return web3
+
+    def is_remote(self) -> bool:
+        return True
+
+    def accounts(self) -> Set[int]:
+        raise NotImplementedError
+
+    def get_nonce(self, address: int) -> int:
+        return self._web3().eth.getTransactionCount(_web3_address(address))
+
+    def get_balance(self, address: int) -> int:
+        return self._web3().eth.getBalance(_web3_address(address))
+
+    def has_storage(self, address: int) -> bool:
+        raise NotImplementedError
+
+    def get_storage(self, address) -> Storage:
+        raise NotImplementedError
+
+    def get_storage_data(self, address: int, offset: Union[int, BitVec]) -> int:
+        if not isinstance(offset, int):
+            raise NotImplementedError
+        return int.from_bytes(self._web3().eth.getStorageAt(_web3_address(address), offset), "big")
+
+    def get_code(self, address: int) -> bytes:
+        return self._web3().eth.getCode(_web3_address(address))
+
+    def get_blocknumber(self) -> int:
+        return self._web3().eth.blockNumber
```

Comment:
I wonder if we should check somewhere that the blocknumber/fork is supported by us.

Reply:
I don't think that should be hard to add. What specifically should be the test?

Reply:
Not sure if it is a valid concern. But when blockNumber is <9999 the transactions are supposed to do one thing and that thing changes with the block. 
If the block is pre istanbul the results may no be reproducible?

No idea how this works in test nets.

Reply:
I mean, what should the check be specifically?  Maybe generate a warning if the blocknumber is less than 4370000, to be consistent with this?
https://github.com/trailofbits/manticore/blob/88ddfbca9efcaa47f736183bee4fd4c6d3ef9b66/manticore/platforms/evm_world_state.py#L169-L171